File: week1/LastStoneWeight/lastStoneWeight.cpp
#include <iostream>
#include <queue>
using namespace std;

int lastStoneWeight(vector<int> stones);

int main(){
    vector<int> input = {2,7,4,1,8,1};
    int output = lastStoneWeight(input);
    printf("last stone: %d", output);
}

int lastStoneWeight(vector<int> stones){
    priority_queue<int> st (stones.begin(), stones.end());

    while (st.size() > 1) {
        int x = st.top();
        st.pop();
        int y = st.top();
        st.pop();
        if (x!=y){
            st.push(abs(x-y));
        }
    }
    
    return st.top();
}

File: week1/LastStoneWeight/lastStoneWeight.py
import heapq

class Solution:
    @classmethod
    def lastStoneWeight(self, stones: list[int])->int:
        heapq._heapify_max(stones)
        print(stones)

        while len(stones) > 1:
            x = heapq.heappop(stones)
            print('x: %d',x)
            y = heapq.heappop(stones)
            print('y: %d',y)
            if(x!=y):
                heapq.heappush(stones, abs(x-y))

        if len(stones) == 0:
            return 0
        else:
            return heapq.heappop(stones)

stones = [2,7,4,1,8,1]
output = Solution.lastStoneWeight(stones)
print(output)